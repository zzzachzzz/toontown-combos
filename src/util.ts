import type { FindComboArgs } from './gags';
import type { GagTrack } from './constants';

export function* batch<T>(batchSize: number, iterable: Iterable<T>): Generator<Array<T>> {
  let yieldNext: Array<T> = [];
  for (const item of iterable) {
    yieldNext.push(item);
    if (yieldNext.length === batchSize) {
      yield yieldNext;
      yieldNext = [];
    }
  }
  if (yieldNext.length > 0)
    yield yieldNext;
}

export function* range(start: number, end: number, step: number = 1): Generator<number> {
  if (step === 0) {
    throw new Error("Step cannot be zero.");
  }
  if (start < end && step < 0 || start > end && step > 0) {
    throw new Error("Invalid range: step direction does not match range direction.");
  }
  if (step > 0) {
    for (let i = start; i <= end; i += step) {
      yield i;
    }
  } else {
    for (let i = start; i >= end; i += step) {
      yield i;
    }
  }
}

/** Generates commbo args for default combos used in UI combos grid */
export function* genFindComboArgsDefault({
  maxCogLvl,
  organicGags,
  isLured,
}: { maxCogLvl: number; } & Pick<
  FindComboArgs,
  'organicGags' | 'isLured'>
): Generator<FindComboArgs> {
  const gagTracks: Array<GagTrack> = ['sound', 'throw', 'squirt', 'drop'];
  const stunTracks: Array<GagTrack> = ['sound', 'throw', 'squirt'];
  for (let cogLvl = maxCogLvl; cogLvl >= 1; cogLvl--) {
    for (const gagTrack of gagTracks) {
      if (gagTrack === 'drop') {
        for (const stunTrack of stunTracks) {
          for (let numToons = 4; numToons >= 1; numToons--) {
            const numDrop = Math.max(numToons - 1, 1);
            yield {
              cogLvl,
              isLured,
              gags: {
                drop: numDrop,
                [stunTrack]: numDrop === numToons ? 0 : 1,
              },
              organicGags,
            };
          }
        }
      } else {
        for (let numToons = 4; numToons >= 1; numToons--) {
          yield {
            cogLvl,
            isLured,
            gags: { [gagTrack]: numToons },
            organicGags,
          };
        }
      }
    }
  }
}

export function* genFindComboArgsCustom({
  gagTrack,
  maxCogLvl = 20,
  minCogLvl = 1,
  maxToons = 4,
  minToons = 1,
  organicGags = {},
  isLured = false,
}: {
  gagTrack: GagTrack;
  maxCogLvl?: number;
  minCogLvl?: number;
  maxToons?: number;
  minToons?: number,
  organicGags?: FindComboArgs['organicGags'];
  isLured?: FindComboArgs['isLured'];
}): Generator<FindComboArgs> {
  for (let cogLvl = maxCogLvl; cogLvl >= minCogLvl; cogLvl--) {
    for (let numToons = maxToons; numToons >= minToons; numToons--) {
      yield { cogLvl, isLured, gags: { [gagTrack]: numToons }, organicGags };
    }
  }
}

export function* genFindComboArgsFixedNumToons({
  maxCogLvl = 20,
  minCogLvl = 1,
  gags,
  organicGags = {},
  isLured = false,
}: {
  maxCogLvl?: number;
  minCogLvl?: number;
  maxToons?: number;
  minToons?: number,
  gags: FindComboArgs['gags'];
  organicGags?: FindComboArgs['organicGags'];
  isLured?: FindComboArgs['isLured'];
}): Generator<FindComboArgs> {
  for (let cogLvl = maxCogLvl; cogLvl >= minCogLvl; cogLvl--) {
    yield { cogLvl, isLured, gags, organicGags };
  }
}

