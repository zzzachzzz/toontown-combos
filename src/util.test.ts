import * as util from './util';

describe('batch', () => {
  test('generates correct batch sizes when remainder', () => {
    const gen = util.batch(4, [1, 2, 3, 4, 5, 6, 7, 8, 9]);
    expect(Array.from(gen)).toEqual([
      [1, 2, 3, 4],
      [5, 6, 7, 8],
      [9],
    ]);
  });

  test('generates correct batch sizes when fewer elements than batchSize', () => {
    const gen = util.batch(6, [1, 2]);
    expect(Array.from(gen)).toEqual([
      [1, 2],
    ]);
  });

  test('generates zero batches when zero elements', () => {
    const gen = util.batch(2, []);
    expect(Array.from(gen)).toEqual([
    ]);
  });
});

describe('iterFindComboArgs', () => {
  test('generated combo args match snapshot', () => {
    const maxCogLvl = 12;
    const organicGags = {};
    const isLured = false;

    expect.addSnapshotSerializer({
      serialize(val, config, indentation, depth, refs, printer) {
        // Output all object properties on one line
        return JSON.stringify(val);
      },
      test(val) {
        return typeof val == 'object' && !Array.isArray(val);
      },
    })
    expect(
      Array.from(util.genFindComboArgsDefault({ maxCogLvl, organicGags, isLured }))
    ).toMatchSnapshot();
  });
});

