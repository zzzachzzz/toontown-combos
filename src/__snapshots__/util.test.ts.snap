// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`iterFindComboArgs > generated combo args match snapshot 1`] = `
[
  {"cogLvl":12,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":12,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":11,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":10,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":9,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":8,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":7,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":6,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":5,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":4,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":3,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":2,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"sound":4},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"sound":3},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"sound":2},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"sound":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"throw":4},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"throw":3},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"throw":2},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"throw":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"squirt":4},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"squirt":3},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"squirt":2},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"squirt":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":3,"sound":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":2,"sound":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":1,"sound":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":1,"sound":0},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":3,"throw":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":2,"throw":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":1,"throw":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":1,"throw":0},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":3,"squirt":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":2,"squirt":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":1,"squirt":1},"organicGags":{}},
  {"cogLvl":1,"isLured":false,"gags":{"drop":1,"squirt":0},"organicGags":{}},
]
`;
